### ISSUE TITLE
Command dispatch

### ISSUE THREAD

**Issue by greenbigfrog** [greenbigfrog/discordtipbot#35]
@z64 mind elaborating where there is still room for improvement?

**Comment by z64** [comment]
As mentioned, this is just my opinion based on my experience coding bots.

`case`-based dispatch works, but of course we both come from backgrounds where you just code the single command, contained, and everything just works, there's nothing else to update. I'm talking of course of things like [gemstone](https://github.com/z64/gemstone), where you just add a file, write up the new container, and it is just included in the bot's handlers.

Since there of course is no container facility for discordcr, most people I've seen end up doing something like this and deferring common behaviors to methods. Here in this bot, you basically have commands-as-methods.

I've done bots in both of these styles;
- DAH, containers https://github.com/z64/discord-against-humanity
- Dex, lots of method-based behavior https://github.com/y32/dex

So, that kind of forms the basis of my opinion:
- I don't want to have to update dispatch logic to include new commands
- Coding a bunch of random helper methods can easily lead to inconsistent / messy implementation if you're not careful. I.e., methods that have similar or pieces of behavior from other methods, or methods that have too much behavior.
- This style also lends it difficult to spec behavior.

Not being happy with the above approaches is what lent me to writing things like [dcrmw](https://github.com/z64/discordcr-middleware) that addresses each of those points in a way I like:

- Creating new "stacks" is opaque with just creating new message event handlers. It feels extremely similar at the moment to discordrb's command bot (stacks are stored exactly the same way, at the moment). There's no dispatch logic to add, each stack stands on its own.
- Code is kept DRY by creating these *self-contained* `Middleware` that are executed in chains (`Stack`s). All the behavior related to that specific "plugin" is contained with it. They are classes themselves, so you can use mixins to share behavior to multiple middleware. Lastly each middleware you make can be used on any stack in any order, and the chain is lazily executed (if one middleware aborts early, the later ones are never called)
- They are easy to spec since they are not tangled into a bot. I've written [a ton of specs](https://github.com/z64/discordcr-middleware/tree/master/spec/middleware) for the set of [stock middleware](https://github.com/z64/discordcr-middleware/tree/master/src/discordcr-middleware/middleware) that I provide.

In short: It gives you kind of a more concrete approach in adding new bot features, i.e. as a programmer you think "Okay, I need a `Middleware` that does XYZ to chain into the next one" rather than just adding methods anywhere that add the behavior you need. There's nothing that forces you to be consistent.

I'm not trying to convince you or anything to port to dcrmw, this is just to illustrate my opinion in what I personally strive for after being spoiled by higher level frameworks like discordrb. If you want to make your own approach to this, I think at the least steps should be made to DRY up / formalize how commands are defined and handled, like a `Commands` mixin, just as we made the `Utility` mixin.

**Comment by greenbigfrog** [comment]
Skimmed through your message. Will have to take a closer look at some later point.
I think this is one of the less relevant "issues" though for now (as you already said yourself).

**Comment by greenbigfrog** [comment]
Will be implemented in #110
